sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/processor/CompactBlockProcessor.kt

```diff
@@ -301,149 +370,396 @@ class CompactBlockProcessor internal constructor(
+        }
 
-        // Syncing last blocks and enhancing transactions
-        var syncResult: BlockProcessingResult = BlockProcessingResult.Success
-        runSyncingAndEnhancing(
+        // This step covers these operations fetchLatestBlockHeight, updateChainTip, suggestScanRanges, updateRange,
+        // and shouldVerifySuggestedScanRanges
+        val preparationResult = runSbSSyncingPreparation(
             backend = backend,
             downloader = downloader,
-            repository = repository,
             network = network,
-            syncRange = syncRange,
-            withDownload = withDownload,
-            enhanceStartHeight = enhanceStartHeight
-        ).collect { syncProgress ->
-            _progress.value = syncProgress.percentage
-            updateProgress(lastSyncedHeight = syncProgress.lastSyncedHeight)
-
-            if (syncProgress.result == BlockProcessingResult.UpdateBirthday) {
-                updateBirthdayHeight()
-            } else if (syncProgress.result != BlockProcessingResult.Success) {
-                syncResult = syncProgress.result
-                return@collect
+            lastValidHeight = lastValidHeight
+        )
+        when (preparationResult) {
+            is SbSPreparationResult.ProcessFailure -> {
+                return preparationResult.toBlockProcessingResult()
+            }
+            SbSPreparationResult.ConnectionFailure -> {
+                return BlockProcessingResult.Reconnecting
+            }
+            SbSPreparationResult.NoMoreBlocksToProcess -> {
+                return BlockProcessingResult.NoBlocksToProcess
+            }
+            is SbSPreparationResult.Success -> {
+                Twig.info { "Preparation phase done with result: $preparationResult" }
+                // Continue processing ranges
             }
         }
 
-        if (syncResult != BlockProcessingResult.Success) {
-            // Remove persisted but not validated and scanned blocks in case of any failure
-            val lastScannedHeight = getLastScannedHeight(repository)
-            downloader.rewindToHeight(lastScannedHeight)
+        var verifyRangeResult = preparationResult.verifyRangeResult
+        var suggestedRangesResult = preparationResult.suggestedRangesResult
+        val lastPreparationTime = System.currentTimeMillis()
+
+        // Running synchronization for the [ScanRange.SuggestScanRangePriority.Verify] range
+        while (verifyRangeResult is VerifySuggestedScanRange.ShouldVerify) {
+            Twig.info { "Starting verification of range: $verifyRangeResult" }
+
+            // Remove existing blocks as they'll be re-downloaded
+            downloader.rewindToHeight(verifyRangeResult.scanRange.range.start)
             deleteAllBlockFiles(
                 downloader = downloader,
-                lastKnownHeight = lastScannedHeight
+                lastKnownHeight = verifyRangeResult.scanRange.range.start
             )
```

Comment:
This is non-atomic; does that matter? I.e. what happens if we rewind but the process is terminated before deleting some or all of the block files?

Reply:
In this scenario, the remaining files will be deleted in the next synchronization attempt in about 30 seconds in most cases or in the next client app start. Both of these cases are fine, as the helper block files are rewritten in the device storage, and they are used only as links to the rust layer to avoid passing all block files to the database.